### PR TITLE
Add batching support to OSS

### PR DESCRIPTION
**Pull request by darwin67**
## Description

The batching feature is current only available on the Cloud.
This change attempts to bring that into OSS and DevServer so users are able to test batching locally.

### Note

This is mainly the backend changes. Showing the batch events on function runs will be a follow up.

## Type of change (choose one)
- [x] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [ ] I've linked any associated issues to this PR.
- [x] I've tested my own changes.

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

### CODE REVIEW

File: pkg/coreapi/gql.schema.graphql [hunk 1]
```diff
@@ -183,6 +183,8 @@ type FunctionRun {
   function: Function
   workspace: Workspace
   event: Event
+  events: [Event]!
+  batchID: ULID # if the run is a batched run
```
Comment (darwin67) [inline pkg/coreapi/gql.schema.graphql:187]: I'll add this resolver as a follow up. Not necessarily needed right now.

File: pkg/consts/consts.go [hunk 2]
```diff
@@ -52,6 +52,9 @@ const (
 	// MaxConcurrencyLimits limits the max concurrency constraints for a specific function.
 	MaxConcurrencyLimits = 2
 
+	// MaxBatchTTL represents the maximum amount of duration the batch key will last
+	MaxBatchTTL = 10 * time.Minute
```
Comment (goodoldneon) [inline pkg/consts/consts.go:56]: The docs say the max batch timeout is 1 minute. Should this value match that?

Comment (darwin67) [reply pkg/consts/consts.go:56]: This is the redis TTL used for expiring the keys after batch started.

File: pkg/coreapi/gql.schema.graphql [hunk 3]
```diff
@@ -183,6 +183,8 @@ type FunctionRun {
   function: Function
   workspace: Workspace
   event: Event
+  events: [Event]!
```
Comment (goodoldneon) [inline pkg/coreapi/gql.schema.graphql:186]:
```suggestion
  events: [Event!]!
```

This change reflects the fact that none of the array elements can be null

File: pkg/devserver/lifecycle.go [hunk 4]
```diff
@@ -33,4 +33,20 @@ func (l lifecycle) OnFunctionScheduled(
...
+			cqrs.WithEventBatchAccountID(id.AccountID),
+			cqrs.WithEventBatchWorkspaceID(id.WorkspaceID),
+			cqrs.WithEventBatchAppID(id.AppID),
+			cqrs.WithEventBatchFunctionID(id.WorkflowID),
+			cqrs.WithEventBatchRunID(id.RunID),
+			cqrs.WithEventBatchEventIDs(id.EventIDs),
+		)
+
+		if batch.IsMulti() {
+			_ = l.cqrs.InsertEventBatch(ctx, *batch)
```
Comment (tonyhb) [inline pkg/devserver/lifecycle.go:49]: This looks good.  I wonder if we end up making batching a lifecycle and then move InsertEventBatch to another interface, which lets us normalize dev server & cloud.  Obvs not a blocker and would be a followup

File: pkg/execution/batch/lua/helpers.lua [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+local function is_status_empty(key)
+  return is_meta_empty(key, "status")
+end
+
+local function set_batch_status(key, value)
+  redis.call("HSET", key, "status", value)
+end
+
+local function get_batch_status(key)
+  return redis.call("HGET", key, "status")
```
Comment (tonyhb) [inline pkg/execution/batch/lua/helpers.lua:29]: i really like making redis calls functions for clarity!

**Review by tonyhb — APPROVED** [review]

File: pkg/devserver/lifecycle.go (hunk 4 above)
Comment (darwin67) [reply pkg/devserver/lifecycle.go:49]: sounds like a plan

**Pull request merged**